doc/explanation/package-management.md

```diff
@@ -0,0 +1,220 @@
+
+Dune defines two repositories by default:
+
+* `upstream` which refers to the default branch of `opam-repository`, which
+  contains all the publically released packages.
+* `overlay` which refers to
+  [opam-overlay](https://github.com/ocaml-dune/opam-overlays), which defines
+  packages patched to work with package management. The long-term goal is to
+  have as few packages as possible in this repository as more and more packages
+  work within Dune package management upstream. Check the
+  [compatibility](#compatibility) section for details.
+
+### Solving
+
+After Dune has retrieved the constraints and the set of possible packages it is
+necessary to determine which packages & versions should be selected for the
+package lock.
+
+To do so, Dune uses
+[opam-0install-solver](https://github.com/ocaml-opam/opam-0install-solver),
+which is a variant of the 0install solver to find solutions for OPAM packages.
+
+Contrary to `opam`, the Dune solver always starts from a blank slate. It
+assumes nothing is installed and everything needs to be installed. This has the
+advantage that solving is simpler and previous solver solutions don't interfere
+with the current one. Thus given the same inputs it should always have the same
+outputs; no state is held between the solver runs.
+
+This can lead to more packages being installed (as OPAM won't by default
+install new versions of packages if the existing versions satisfy the
+constraints) but it avoids interference from already installed packages that
+lead to potentially different solutions.
+
+After solving is done, the solution gets written into the lock directory with
+all the data that is necessary to build and install the packages. From this
+point, the project contains all necessary information and does not need to
+access any package repositories.
+
+## Building
+
+When building, Dune will read the information from the lock directory and set
+up build rules to use the packages. The rules include
+
+* Fetch rules to download the sources as well as any additional sources like
+  patches and unpack them
+* Build rules to evaluate the build instructions from the build instructions
+  stored in the lock directory
+* Install rules to put the artifacts that were built into the appropriate,
+  Dune-managed folders
+
+Creating these processes as rules means that they will only be executed on
+demand, thus if the project has already downloaded the sources, it does not
+need to download them again. Likewise, if packages are installed, they stay
+installed.
+
+The results of the rules are stored in the `_build` directory of the project
+and managed automatically by Dune. Thus when cleaning the build directory, the
+installed packages are cleaned as well and will be reinstalled at the next
+build.
+
+When building the users project, the installed packages are added to the
+necessary search paths, thus user code can use the dependencies without any
+additional configuration.
+
+(compatibility)=
```

Comment:
```suggestion
```
Is it Sphinx syntax or a typo?

Reply:
It's MyST (the markdown plugin for reST/RTD) syntax, so I can refer to that section as `#compatibility` elsewhere. There's a whole stack of Markdown, MyST and RTD things overlapping so determining what syntax to use can be a bit surprising.